flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinatorTriggeringTest.java

```diff
@@ -529,6 +537,48 @@ public void testTriggerCheckpointSnapshotMasterHookFailed() throws Exception {
 		assertEquals(0, checkpointCoordinator.getTriggerRequestQueue().size());
 	}
 
+	/**
+	 * This test only fails eventually.
+	 */
+	@Test
+	public void discardingTriggeringCheckpointWillExecuteNextCheckpointRequest() throws Exception {
+		final ExecutionVertex executionVertex = mockExecutionVertex(new ExecutionAttemptID());
+
+		final ScheduledExecutorService scheduledExecutorService = Executors.newSingleThreadScheduledExecutor();
+		final CheckpointCoordinator checkpointCoordinator = new CheckpointCoordinatorTestingUtils.CheckpointCoordinatorBuilder()
+			.setTasks(new ExecutionVertex[]{executionVertex})
+			.setTimer(new ScheduledExecutorServiceAdapter(scheduledExecutorService))
+			.setCheckpointCoordinatorConfiguration(CheckpointCoordinatorConfiguration.builder()
+				.build())
+			.build();
+
+		final CompletableFuture<String> masterHookCheckpointFuture = new CompletableFuture<>();
+		final OneShotLatch triggerCheckpointLatch = new OneShotLatch();
+		checkpointCoordinator.addMasterHook(new TestingMasterHook(masterHookCheckpointFuture, triggerCheckpointLatch));
+
+		try {
+			checkpointCoordinator.triggerCheckpoint(false);
+			final CompletableFuture<CompletedCheckpoint> secondCheckpoint = checkpointCoordinator.triggerCheckpoint(false);
+
+			triggerCheckpointLatch.await();
+			masterHookCheckpointFuture.complete("Completed");
+
+			// discard triggering checkpoint
+			checkpointCoordinator.abortPendingCheckpoints(new CheckpointException(CheckpointFailureReason.CHECKPOINT_DECLINED));
+
+			try {
+				// verify that the second checkpoint request will be executed and eventually times out
+				secondCheckpoint.get();
```

Comment:
I guess I'm missing some changes because I see only one checkpoint in this test.

I think there indeed should be a second checkpoint that is started after the previous one is aborted. We can simulate the original problem without a latch by completing `TestingMasterHook.checkpointFuture` after aborting the checkpoint.

Reply:
The first checkpoint is the one we are aborting. The latch is needed because we don't want to abort the first checkpoint too fast before it hasn't been created yet.

Reply:
Strictly speaking it would also work w/o the latch but it increases the likelihood of the right thread interleaving here.

Reply:
> The first checkpoint is the one we are aborting

Oh, I see :)

I ran this test against master and it passed. I played a bit with it but didn't manage to make it fail.

Reply:
For me it failed when running it in the IDE with until failure.

Reply:
Then it should probably be in a loop.

Reply:
It couldn't fail on my side either (based on release-1.11). I think the reason is there is no asynchronous call completed after the pending checkpoint has been aborted.

I would suggest to use `manuallyTriggeredScheduledExecutor` as timer thread and use a separate thread for IO executor. Because,
1. It's closer to the real scenario and we can control the sequence of each invocation accurately. We can easily abort the pending checkpoint first and complete an IO asynchronous invocation to simulate the NPE scenario.
2. In the later PRs, when main thread is introduced to replace timer thread (now it's reverted due to FLINK-16770, I'm preparing to get them back), the coordinator-wide lock would be removed, 
`triggerCheckpoint` and `abortPendingCheckpoints` would be called only in main thread. At that time, we have to use `manuallyTriggeredScheduledExecutor` as main thread executor to simulate the lock-free and single-threaded threading model. If we do it here, we don't need change it then.

Reply:
The test case should fail eventually. It is in the nature of concurrency related bugs that sometimes they require a certain thread interleaving which in this case is not easy to reproduce. I will check whether the test fails on my machine.

Concerning the `manuallyTriggeredSchedulerExecutor`, I deliberately did not use it because it leaks internal implementation details. Concretely, the test has to know how many tasks are being enqueued until one reaches the `whenCompleteAsync` call. I think this is very brittle and should be avoided if possible. Imagine that someone changes the implementation of the `CheckpointCoordinator` so that there is another stage scheduled, this would immediately break the test w/o a very visible reason.

Reply:
The test fails on my machine w/o the fix. Sometimes it needs more than 3000 runs though. Not sure how long you have it let run.